### ISSUE TITLE
OOMKilled with BULK INSERT

### ISSUE THREAD

**Issue by reversTeam** [MaterializeInc/materialize#5611]
### What version of Materialize are you using?

v0.6.1 from docker image

### How did you install Materialize?

* [x] Docker image
* [ ] Linux release tarball
* [ ] APT package
* [ ] macOS release tarball
* [ ] Homebrew tap
* [ ] Built from source

### What was the issue?

Today I try to materialize to process large volumes of data, I go through BULK INSERT (by batch of 500 items) in order to ingest the data.
 - CSV: not retained, because the process is detached from the database pod
 - Kafka Stream / Proto: Not possible, for internal reasons
 - ...

What I observe at the level of the database pod is that the RAM is growing in an unlimited way.

I have changed the compaction config variables, from 60 to 10 then to 1s, but I don't see any change in memory usage.
I wonder if the insert in bulk, does not trigger compaction or if there would be some other underlying cause.

### Is the issue reproducible? If so, please provide reproduction instructions.

Just do BULK INSERT (with materialized view) with a compaction configuration to realize that the database will grow to OOMKilled.

### Please attach any applicable log files.

```
Last State:     Terminated
Reason:       OOMKilled
```

**Comment by cirego** [comment]
Hi @reversTeam, thanks for filing an issue!

Materialize is currently in-memory only. Does the dataset that you're trying to process fit in the memory that you have available on your system? Tweaking compaction may help if you are overwriting data, but if you are simply inserting / appending data, then I think this might be the expected behavior on large datasets.

**Comment by reversTeam** [comment]
Indeed, the tests are carried out with several hundreds of millions of lines. And I had hoped to be able to make materialized views over several tens of billions of lines. With reasonable instance sizing.

Indeed the compaction is configured, I tried several values, 60, 10 and 1sec, but I admit not having seen any difference, that's why I thought that there could be a problem on the BULK INSERT.

Do you have a project under study that could solve the problem?

**Comment by cirego** [comment]
Hi @reversTeam, can you provide more information on the size of the total dataset (10GB, 100GB?) and the instance type that you are running on?

If you are only appending new data, then the compaction parameters should not affect the amount of memory used.

**Comment by cirego** [comment]
Also, if you can provide the materialized view that you are running, that would be helpful in diagnosing this situation.

**Comment by reversTeam** [comment]
Hi @cirego, I am currently working with an instance of 64 cores, + 240GB of ram (220GB dedicated for materialize).
I think I was pus ingest before being OOMKill about 400 Million events.

In projecting the memory footprint to use, we cannot consider materialize as a solution to our problem.
We would speak of a server with several TB of RAM to suit the end use case.

I am currently storing 3 basics informations for a size of 60 bytes per row.
 - x: uuid
 - y: characteristic
 - z: date
 ```
 CREATE TABLE  ingest_table (
    x TEXT NOT NULL,
    y TEXT NOT NULL,
    z TIMESTAMP WITH TIME ZONE NOT NULL
);
 ```
 
I insert via a feed when I receive the type of events I want to save.

I create for this table 6 materialized views:
 - Count distinct uuid per day
 ```
 CREATE MATERIALIZED VIEW table_count_x_per_week
    AS SELECT
        COUNT(DISTINCT(x)) AS x,
        date_trunc('day', z) AS date
    FROM ingest_table
        GROUP BY date
    ORDER BY date;
 ```
 - Count distinct uuid per week
 ```
 CREATE MATERIALIZED VIEW table_count_x_per_week
    AS SELECT
        COUNT(DISTINCT(x)) AS x,
        date_trunc('week', z) AS date
    FROM ingest_table
        GROUP BY date
    ORDER BY date;
 ```
 - Count distinct uuid per month
 ```
 CREATE MATERIALIZED VIEW table_count_x_per_month
    AS SELECT
        COUNT(DISTINCT(x)) AS x,
        date_trunc('month', z) AS date
    FROM ingest_table
        GROUP BY date
    ORDER BY date;
 ```
 
 - Count distinct characteristic per day
 ```
 CREATE MATERIALIZED VIEW table_count_y_per_day
    AS SELECT
        COUNT(DISTINCT(x)) AS x,
        y,
        date_trunc('day', z) AS date
    FROM ingest_table
        GROUP BY date, y
    ORDER BY date;
 ```
 - Count distinct characteristic per week
 ```
 CREATE MATERIALIZED VIEW table_count_y_per_week
    AS SELECT
        COUNT(DISTINCT(x)) AS x,
        y,
        date_trunc('week', z) AS date
    FROM ingest_table
        GROUP BY date, y
    ORDER BY date;
 ```
 - Count distinct characteristic per month
 ```
 CREATE MATERIALIZED VIEW table_count_y_per_month
    AS SELECT
        COUNT(DISTINCT(x)) AS x,
        y,
        date_trunc('month', z) AS date
    FROM ingest_table
        GROUP BY date, y
    ORDER BY date;
 ```
  
  Thank you for your support, the functionality offered by your solutions is brilliant and innovative !

**Comment by wangandi** [comment]
Hi @reversTeam,
The problem is that `CREATE TABLE` is not suitable for large amounts of streaming data, and it doesn't matter what kinds of views you are creating on top of the table. Let me try to provide the intuition for why this is.

As stated in the [`CREATE TABLE` documentation](https://materialize.com/docs/sql/create-table/#main), "CREATE TABLE creates a non-streaming, in-memory data source." When you create a table, Materialize does not know what kinds of views you will create over the table. In the comment above, you list that you have created six Materialize views, but even once you finish creating those views, Materialize does not know whether you will want to create more views. Thus, it cannot throw or compact away any of the rows because a future view might need that row. 

What you're looking for is [`CREATE SOURCE`](https://materialize.com/docs/sql/create-source/). As stated in the docs page, "Sources represent connections to resources outside Materialize that it can read data from." If you create a source and then create materialized views on top of the source, Materialize will only save the data relevant to each materialized view you create. This is because if you create a new materialized view in the future, and the data you need does not exist in Materialize can just go and read that data from the outside resource.

tl; dr. `CREATE TABLE` tells Materialize to treat what is in memory as the source of truth, so Materialize has to save everything. `CREATE SOURCE` tells Materialize that there is a source of truth outside, so Materialize only needs to save what is needed to maintain your views in real time.

If you tell us where what kind of database your data is coming from, we may be able to help in figuring out how to stream that data into Materialize without creating a table.

**Comment by reversTeam** [comment]
Hi @wangandi,

Following some discussion around my implementation.
I have to admit that I haven't made good use of the technology.
Thank's to confirm this, and your explanations are in perfect harmony with our conclusions.

Thank you for these explanations, I will review my implementation to use the stream from a source, as I should have done from the beginning.
